redb-derive/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+    let input = parse_macro_input!(input as DeriveInput);
+
+    match expand_derive_value(input) {
+        Ok(tokens) => tokens.into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+fn expand_derive_value(input: DeriveInput) -> syn::Result<proc_macro2::TokenStream> {
+    let struct_name = &input.ident;
+    let struct_name_str = struct_name.to_string();
+
+    let fields = match &input.data {
+        Data::Struct(data_struct) => match &data_struct.fields {
+            Fields::Named(fields) => &fields.named,
+            Fields::Unnamed(fields) => &fields.unnamed,
+            Fields::Unit => {
+                return Err(syn::Error::new_spanned(
+                    struct_name,
+                    "Unit structs are not supported",
+                ));
+            }
+        },
+        _ => {
+            return Err(syn::Error::new_spanned(
+                struct_name,
+                "Only structs are supported",
+            ));
+        }
+    };
+
+    if fields.is_empty() {
+        return Err(syn::Error::new_spanned(
+            struct_name,
+            "Empty structs are not supported",
+        ));
+    }
+
+    let field_types: Vec<_> = fields.iter().map(|f| &f.ty).collect();
+    let field_count = fields.len();
+
+    let tuple_type = if field_count == 1 {
+        let ty = &field_types[0];
+        quote! { (#ty,) }
+    } else {
+        quote! { (#(#field_types),*) }
+    };
+
+    let is_named_fields = fields.iter().next().unwrap().ident.is_some();
+
+    let (struct_to_tuple, tuple_to_struct) = if is_named_fields {
+        let field_names: Vec<_> = fields.iter().map(|f| f.ident.as_ref().unwrap()).collect();
+        let struct_to_tuple = if field_count == 1 {
+            let field = &field_names[0];
+            quote! { (value.#field.clone(),) }
+        } else {
+            quote! { (#(value.#field_names.clone()),*) }
+        };
+
+        let tuple_to_struct = {
+            let indices: Vec<Index> = (0..field_count).map(Index::from).collect();
+            quote! {
+                #struct_name {
+                    #(#field_names: tuple.#indices.clone()),*
+                }
+            }
+        };
+        (struct_to_tuple, tuple_to_struct)
+    } else {
+        let indices: Vec<Index> = (0..field_count).map(Index::from).collect();
+        let struct_to_tuple = if field_count == 1 {
+            quote! { (value.0.clone(),) }
+        } else {
+            quote! { (#(value.#indices.clone()),*) }
+        };
+
+        let tuple_to_struct = quote! {
+            #struct_name(#(tuple.#indices.clone()),*)
+        };
+        (struct_to_tuple, tuple_to_struct)
+    };
+
+    let type_name_with_fields = if is_named_fields {
+        let field_names: Vec<_> = fields.iter().map(|f| f.ident.as_ref().unwrap()).collect();
+        let field_type_names = field_types.iter().map(|ty| {
+            quote! { stringify!(#ty).replace(" ", "") }
```

Comment:
Instead of stringifying the `#ty`, let's use `format!("{}", #ty::type_name())` to get the redb type_name() implementation.

Also, return a friendly error message if any of the fields do NOT implement `redb::Value`, since our serialization and deserialization rely on that